tests/testthat/test-deterministic.R

```diff
@@ -0,0 +1,313 @@
+  set.seed(1)
+  ll1 <- p1$run(dat$pars$model(dat$pars$initial()),
+                save_history = TRUE)
+
+  compare <- function(state, observed, pars = NULL) {
+    if (is.na(observed$incidence)) {
+      return(NULL)
+    }
+    if (is.null(pars$compare$exp_noise)) {
+      exp_noise <- 1e6
+    } else {
+      exp_noise <- pars$compare$exp_noise
+    }
+    ## This is on the *filtered* state (i.e., returned by run())
+    incidence_modelled <- state["incidence", , drop = TRUE]
+    incidence_observed <- observed$incidence
+    lambda <- incidence_modelled +
+      rexp(n = length(incidence_modelled), rate = exp_noise)
+    dpois(x = incidence_observed, lambda = lambda, log = TRUE)
+  }
+
+  index <- function(info) {
+    list(run = c("S" = 1L, "incidence" = 5L),
+         state = c("S" = 1L, "I" = 2L, "R" = 3L))
+  }
+
+  p2 <- particle_nofilter$new(dat$data, dat$model, compare, index)
+  set.seed(1)
+  ll2 <- p2$run(dat$pars$model(dat$pars$initial()),
+                save_history = TRUE)
+  expect_identical(ll2, ll1)
+  h1 <- p1$history()
+  h2 <- p2$history()
+
+  expect_equal(h1, unname(h2))
+  expect_equal(dimnames(h2), list(c("S", "I", "R"), NULL, NULL))
+})
+
+
+test_that("initial passes args as expected, for multipar case", {
+  ## As normal, but we require the exp_noise parameter
+  compare <- function(state, observed, pars = NULL) {
+    if (is.na(observed$incidence)) {
+      return(NULL)
+    }
+    exp_noise <- pars$exp_noise
+    incidence_modelled <- state[1L, , drop = TRUE]
+    incidence_observed <- observed$incidence
+    lambda <- incidence_modelled +
+      rexp(n = length(incidence_modelled), rate = exp_noise)
+    dpois(x = incidence_observed, lambda = lambda, log = TRUE)
+  }
+
+  dat <- example_sir()
+  p <- particle_nofilter$new(dat$data, dat$model, compare,
+                             index = dat$index, initial = dat$initial)
+  pars <- list(list(exp_noise = Inf),
+               list(exp_noise = 1e6))
+  ll <- p$run_many(pars)
+  expect_equal(ll[[1]], ll[[2]], tolerance = 1e-5)
+  expect_false(identical(ll[[1]], ll[[2]]))
+})
+
+
+test_that("reconstruct deterministic filter from inputs", {
+  dat <- example_sir()
+  p1 <- particle_nofilter$new(dat$data, dat$model, dat$compare, dat$index)
+  inputs <- p1$inputs()
+  p2 <- particle_filter_from_inputs(inputs)
+  expect_s3_class(p2, class(p1))
+  expect_equal(p1$run(), p2$run())
+})
+
+
+test_that("Can run parallel mcmc with deterministic model", {
+  dat <- example_sir()
+  n_steps <- 30L
+  n_chains <- 3L
+  control <- pmcmc_control(n_steps, save_trajectories = FALSE,
+                           n_workers = 2L, n_chains = n_chains)
+  p <- particle_nofilter$new(dat$data, dat$model, dat$compare, dat$index)
+  res <- pmcmc(dat$pars, p, NULL, control)
+  expect_s3_class(res, "mcstate_pmcmc")
+  expect_equal(nrow(res$pars), n_chains * (n_steps + 1))
+})
+
+
+test_that("Can run parallel mcmc with deterministic model", {
```

Comment:
This and the test immediately above have the same description - does one of them need changing?

Reply:
thanks, good spot